src/waiter.c

```diff
@@ -0,0 +1,311 @@
+            waiter, notification_callback, context);
+    }
+    else
+    {
+        SM_RESULT sm_exec_begin_result = sm_exec_begin(waiter->sm);
+        if (sm_exec_begin_result != SM_EXEC_GRANTED)
+        {
+            LogError("Failure in sm_exec_begin(waiter->sm=%p)", waiter->sm);
+        }
+        else
+        {
+            bool success;
+            srw_lock_acquire_exclusive(waiter->lock);
+            {
+                if (waiter->registered)
+                {
+                    LogError("waiter_register_notification called twice on WAITER waiter=%p."
+                    " Previous arguments: NOTIFICATION_CALLBACK notification_callback=%p, void* context=%p"
+                    " Current arguments: NOTIFICATION_CALLBACK notification_callback=%p, void* context=%p",
+                    waiter,
+                    waiter->notification_callback, waiter->notification_callback_context,
+                    notification_callback, context
+                    );
+                    success = false;
+                }
+                else
+                {
+                    THANDLE(ASYNC_OP) async_op = async_op_create(cancel_register_notification, sizeof(WAITER_HANDLE), alignof(WAITER_HANDLE), dispose);
+                    if (async_op == NULL)
+                    {
+                        LogError("Failure in async_op_create(cancel_register_notification=%p, sizeof(WAITER_HANDLE)=%" PRIu32 ", alignof(WAITER_HANDLE)=%" PRIu32 ", dispose=%p)", cancel_register_notification, (uint32_t)sizeof(WAITER_HANDLE), (uint32_t)alignof(WAITER_HANDLE), dispose);
+                        success = false;
+                    }
+                    else
+                    {
+                        ((ASYNC_OP*)async_op)->context = waiter;
+                        waiter->notification_callback = notification_callback;
+                        waiter->notification_callback_context = context;
+                        waiter->registered = true;
+                        try_complete(waiter);
+                        THANDLE_INITIALIZE_MOVE(ASYNC_OP)(&result, &async_op);
+                        success = true;
+                    }
+                }
+            }
+            srw_lock_release_exclusive(waiter->lock);
+
+            if (!success)
+            {
+                LogError("Failure in waiter_register_notification(waiter=%p, notification_callback=%p, context=%p)", waiter, notification_callback, context);
+            }
+            else
+            {
+                goto all_ok;
+            }
+            sm_exec_end(waiter->sm);
+        }
+    }
+all_ok:
+    return result;
+}
+
+static void cancel_notify(void* context)
+{
+    WAITER_HANDLE waiter = context;
+    srw_lock_acquire_exclusive(waiter->lock);
+    {
+        waiter->notify_complete_callback(waiter->notify_complete_callback_context, WAITER_RESULT_CANCELLED);
+        waiter->callbacks_called = true;
```

Comment:
>waiter->callbacks_called = true; [](http://example.com/codeflow?start=8&length=32)

It is funny to see some plural "callback*s*_call" and it is just 1...

Reply:
Yeah it really should be callback(s)_called.